### ISSUE TITLE
RScriptManager

### ISSUE THREAD

**Issue by adavydenko** [lanit-tercom-school/analyzeme#100]
Необходимо создать класс, контролирущий работу R-скриптов.
Требования:

1. Должен быть один инстанс на приложение. Для начала можно сделать в виде Singletone-а или фабрики, которая возвращает один и тот же инстанс
2. Должен поддерживать многопоточную работу
3. Работает между `IAnalyzer `и `RFacade`

Возможный интерфейс:
``` java
pubic interface IScriptManager {
  Token QueueScriptAsync<TScript, TResult> (Func<TResult> callback, TScript scriptToRun, params[]) {}
  Status GetRunningStatus (Token scriptRunId) {}
  CancelResult CancelRunning (Token scriptRunId) {}
  ScriptInfo[] GetScripts () {}
}

```

**Comment by lagroffe** [comment]
_Работает между IAnalyzer и RFacade_
Тогда нужно, наверное, решить, к чему мы будем обращаться из классов с логикой - к RAnalyzerам или к RFacade. RAnalyzerы сами по себе никакой смысловой нагрузки не несут, просто у них интерфейсы поудобнее ([RAnalyzerToGetNumber](https://github.com/lanit-tercom-school/analyzeme/blob/RinJava/src/main/java/com/analyzeme/R/analyze/RAnalyzerToGetNumber.java), например). Или скрипт-менеджер будет как-то контролировать любое обращение к RFacade, без разницы, откуда?

**Comment by adavydenko** [comment]
`ScriptManager` должен управлять запусками скриптов и отслеживать их состояние.
Наверное проще этого достичь, если он сам будет запускать скрипты, нежели мониторить сторонние процессы.

Я пока понимал так, что `RAnalyzer` это есть какой-то класс с логиков, т.е. _класс, который умеет выполнять R_

**Comment by lagroffe** [comment]
По поводу `RAnalyzer` - я имела в виду те сущности, которые сейчас так называются (их действительно есть смысл переименовать, только я не знаю, как). 
Их суть в том, что, т.к. мы решили, что возвращаемое значение знаем заранее, человеку, который будет писать логику обращения к R, возможно легче вызвать их метод со стандартным названием, а не разбираться в многообразии `RFacade`. Но это было сделано чисто на мой вкус, возможно, их стоит вообще убрать и оставить просто фасад. Как раз из-за этого и был вопрос - мониторинг возникнет перед ними (что удобнее использовать, их или фасад?) или непосредственно перед фасадом?

**Comment by lagroffe** [comment]
И ещё по поводу CancelRunning - для средств интеграции пытаться искать какой-то механизм, который позволяет прервать их работу во время выполнения команды, или это будет просто игнорирование результата?

**Comment by lagroffe** [comment]
К тому, что выше - гипотетически, если это не противоречит ничему существенному, вполне логично было бы сам фасад возможностью запоминать и контролировать то, что он делает, и наделить. Сейчас его действия заключаются исключительно в выборе объекта интеграции, которому он передаёт данные в том же виде, что и получил. 
Тогда и то, что нужный по этому issue класс _должен управлять запусками скриптов и отслеживать их состояние_ будет более чем выполнятся, и лишние классы плодиться не будут - потому что новому ScriptManager, если он будет передавать данные от анализатора фасаду, всё равно придётся повторить абсолютно все его функции практически в том же виде - за исключением хранения и создания одного объекта

**Comment by adavydenko** [comment]
`CancelRunning` - приведен пока чисто для примера, как это реализовывать пока не знаю. Можно через `throw new NotImplementedException()`. В действительности реализация будет сильно зависеть от того, кто этот процесс хостит, возможно обработчик R будет предоставлять такие средства из коробки, ..а может быть и нет..

Идею перенести мониторинг в `RFacade` не поддерживаю, т.к. не хочется чрезмерно усложнять класс. Фасад предназначен для сокрытия сложности цепочки вызовов от пользователя, своего рода инкапсуляция логики. А мониторинг - отдельно независимая задача.
По поводу где его проводить - за, или перед фасадом: если мы предполагаем, что у нас возможны разные фасады (гипотетически) к разной инфрастуктуре. Или еще хуже, если запросы перенаправляются на разные инстансы (компьютеров), то тогда `ScriptManager` должен находиться перед ними. В противном случае, количество менеджеров будет равно количеству физических фасадов на которых производится обработка.
//На своем мнении не настаиваю, если проще сделать по-другому - давайте делать.

**Comment by lagroffe** [comment]
Ага, тогда на будущее.

Хорошо, если они будут отдельными, тогда `ScriptManager`, видимо, нужно будет располагать перед `RFacade`. `RFacade` от инфрастуктуры не должен измениться в любом случае ([вот он](https://github.com/lanit-tercom-school/analyzeme/blob/RinJava/src/main/java/com/analyzeme/R/facade/RFacade.java), если что), но ему для обращения к интеграции будут нужны настройки - а их, если я правильно понимаю, в будущем возможно будет изменять `ScriptManager` (например, для более оптимального использования инстансов Rserve, отмечая, кто из них не занят). 
Тогда у меня остаётся только вопрос по поводу классов, которые сейчас `RAnalyzer` (ссылка на пример есть выше в комментариях). Их оставлять? При появлении `ScriptManager` перед фасадом в них появляется более очевидный смысл - без них скрипты будут передаваться `ScriptManager`, а команды - `RFacade`, лишняя путаница. А так нужно будет изменить их функции для скриптов на обращение не к фасаду, а к `ScriptManager`, а из внешних, настоящих `IAnalyzer` вызывать их, не думая ни о менеджере, ни о фасаде, ни тем более об интеграции.
(и как их, RAnalyzer, лучше назвать, случайно нет предложений?)

**Comment by adavydenko** [comment]
Наверное я правильно понял вашу идею. Попробую подэтожить:
 - `IAnalyzer` - это _внешний_ интерфейс для пользователя, чтобы провести какой-то анализ над данными (напр. посчитать _GlobalMinMax_). Если пользователь хочет `R`, то это будет делать специальный ~`RAnalyzer`
 - `ScriptManager` - умеет выполнять R срипты. Как-то, как знает только он. Для этого внутри он использует `Facade`
 - `RFacade` умеет физически организовывать выполнение R на определенной инфраструктуре
 - `R.Caller` (или как он там называется) - умеет запускать R на Rserve/JRI/rinjin

Верно?

**Comment by lagroffe** [comment]
По-моему да, так. Только по поводу `RAnalyzer` (дальше `GetSthFromR`) распишу, как я понимаю его функции, чтобы если есть возражения, можно было подкорректировать.
`IAnalyzer`, если ему нужен R, обращается к общей обёртке для `ScriptManager` (для скриптов) и `RFacade` (для одиночных команд, если для них тоже не появится какой-то мониторинг - но это будет проблемой обёртки). Эта обёртка прячет, что это разные сущности, и предоставляет интерфейс только для одного возвращаемого значения. Т.е. пользователь в `IAnalyzer` знает, что он хочет вызвать R и получить double, он вызывает runScript или runCommand у какого-то класса `getDoubleFromR`(или, возможно, `getFromR<Double>`). Это сильно уменьшит интерфейсы, которые нужно просмотреть, чтобы найти нужную команду, и пользователю будет всё равно, есть там какой-то монитор скриптов или они просто передаются `RFacade`

**Comment by adavydenko** [comment]
Небольшая корректировка, под `IAnalyzer` я понимаю то, что отображается у нас в комбобоксе, когда пользователь хочет вызвать функции.

**Comment by lagroffe** [comment]
Да, или что-то прямо в Java, или, в случае, связанном с этим issue, вызов заранее заданной команды/скрипта. 
А когда будет редактор для скриптов/возможность задавать команды самостоятельно, из его контроллера (или каких-то дополнительных классов) обращение тоже будет к `GetSthFromR`. Так?

**Comment by adavydenko** [comment]
"внутри него будет или что-то прямо в Java, или, в случае, связанном с этим issue, вызов заранее (нами) заданной команды/скрипта" - не совсем, я думаю расширение функциональности нужно делать через раследование. Т.е.:
```java
interface IAnalyzer {}

public class GlobalMinAnalyzer: IAnalyzer {}
public class GlobalMaxAnalyzer: IAnalyzer {}
public class LyapunovAnalyzer: IAnalyzer {}
public class RAnalyzer: IAnalyzer {}
/*
and other great analyzers..
*/
```

**Comment by lagroffe** [comment]
Ну да, я это и имела в виду - внутри классов, которые его имплементируют, внутри него же всё равно ничего нет. 
Но в случае с `RAnalyzer` (я понимаю, что это пример, но он существенно отличается от трёх соседних сутью названия, хочу понять, что под ним могло бы иметься в виду) - это, в этой трактовке, будет как раз тот класс, который будет отвечать за написанные пользователем скрипты, или что? Потому что теоретически никто не запрещает обращаться к R из `GlobalMinAnalyzer` или любого другого - если тому, кто его пишет, не захотелось делать это на Java. А для только что написанных пользователем скриптов удобнее разработать другой интерфейс, потому что у них будет другой контроллер и другие данные, передаваемые от пользователя.

**Comment by adavydenko** [comment]
Я думал, что да, будет отвечать за написанные пользователем скрипты.

**Comment by lagroffe** [comment]
Хорошо, тогда идея понятна, а как будет удобнее это делать, посмотрим по ходу дела

**Comment by lagroffe** [comment]
При создании RScriptManager должен сохранить уже реализованные [функции](https://github.com/lanit-tercom-school/analyzeme/blob/RinJava/src/main/java/com/analyzeme/R/ScriptManager/RScriptManager.java), дополнив их любой нужной ему логикой
